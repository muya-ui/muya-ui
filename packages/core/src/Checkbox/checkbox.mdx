---
name: Checkbox 复选框
menu: 组件
route: /components/checkbox
autoDemos: true
parent: 数据录入
---

import { DocCheckboxStyle } from './demos/styled';

<DocCheckboxStyle />

# Checkbox 复选框

在可选项中进行多项选择时使用，拥有两种状态切换的基础组件。

## 使用场景

- 在一组可选项中进行多项选择时。
- 单独使用可以表示两种状态之间的切换，和 switch 类似。

## 组件

- `Checkbox`: 选择器。
- `CheckboxGroup`: 选择器组。

{Demos}

### Hooks

#### useCheckbox

Hooks 内部使用以下五个参数：`checked`, `defaultChecked`, `disabled`, `onChange`, `onClick`。

返回参数如下：

- checkedState：状态
- handleChange、handleClick：DOM 事件
- setChecked: 手动更改状态函数

使用范例：

```jsx
const CustomCheckBox = (props: ICheckboxProps) => {
  const { checkedState, handleClick } = useCheckbox(props, 'checked' in props);
  return (
    <span
      style={{
        backgroundColor: checkedState ? '#1a7af8' : 'rgb(248, 248, 248)',
        color: checkedState ? 'white' : 'inherit',
      }}
      onClick={handleClick}
    >
      {props.children}
    </span>
  );
};
const Demo = (props: ICheckboxGroupProps) => {
  return (
    <CustomCheckBox
      key={String(value)}
      disabled={disabled}
      checked={valueState.indexOf(value) > -1}
      onClick={handleClick.bind(null, option)}
    >
      Demo
    </CustomCheckBox>
  );
};
```

#### useCheckboxGroup

函数参数类型同 `ICheckboxGroupProps`。

使用范例：

```jsx
const CustomCheckBox = (props: ICheckboxProps) => {
  const { checkedState, handleClick } = useCheckbox(props, 'checked' in props);
  return (
    <span
      style={{
        backgroundColor: checkedState ? '#1a7af8' : 'rgb(248, 248, 248)',
        color: checkedState ? 'white' : 'inherit',
      }}
      onClick={handleClick}
    >
      {props.children}
    </span>
  );
};

const CustomCheckBoxGroup = (props: ICheckboxGroupProps) => {
  const { opts, valueState, handleClick, restProps } = useCheckboxGroup(props);
  return (
    <div
      {...restProps}
      style={{
        display: 'flex',
        marginBottom: '10px',
      }}
    >
      {opts.map(option => {
        const { label, value, disabled } = option;
        return (
          <CustomCheckBox
            key={String(value)}
            value={value}
            disabled={disabled}
            checked={valueState.indexOf(value) > -1}
            onClick={handleClick.bind(null, option)}
          >
            {label}
          </CustomCheckBox>
        );
      })}
    </div>
  );
};

function GroupDemo() {
  const plainOptions: string[] = ['全部', '审核中', '已通过'];
  const [value, setValue] = useState<ICheckboxGroupValue[]>([plainOptions[1]]);

  const onChange = (checkedValue: ICheckboxGroupValue[]) => {
    console.log('checked: ', checkedValue);
    setValue(checkedValue);
  };

  return (
    <div>
      <CustomCheckBoxGroup
        value={value}
        options={plainOptions}
        onChange={onChange}
      />
    </div>
  );
}
```

### API

#### ICheckboxProps

<Props propsName="ICheckboxProps" />

#### ICheckboxGroupProps

ICheckboxGroupValue：`string | number | boolean`

<Props propsName="ICheckboxGroupProps" />

#### ICheckboxGroupOption

<Props propsName="ICheckboxGroupOption" />

### Styles

#### ICheckboxStyleKeys

<Styles typeName="ICheckboxStyleKeys" componentName="checkbox" />
