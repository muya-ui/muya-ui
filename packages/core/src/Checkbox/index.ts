export { default } from './Checkbox';

export { default as Checkbox } from './Checkbox';
export { default as CheckboxGroup } from './CheckboxGroup';
export { default as useCheckbox } from './useCheckbox'
export { default as useCheckboxGroup } from './useCheckboxGroup'
export * from './types';
