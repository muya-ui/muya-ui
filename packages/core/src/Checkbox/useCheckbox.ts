import React, { useState, useCallback } from 'react';

import { ICheckboxBaseProps } from './types';
import { Omit } from '../types';

type Props = Omit<ICheckboxBaseProps, 'styles'>;

export default function useCheckbox<T extends HTMLElement>(props: Props, isControlled: boolean) {
  const { checked, defaultChecked, disabled, onChange, onClick } = props;
  const [checkedState, setChecked] = useState((isControlled ? checked : defaultChecked) || false);

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      if (disabled) {
        return;
      }
      if (!isControlled) {
        setChecked(e.target.checked);
      }
      if (onChange) {
        onChange(e);
      }
    },
    [disabled, isControlled, onChange],
  );

  const handleClick = useCallback(
    (e: React.MouseEvent<T, MouseEvent>) => {
      if (disabled) {
        return;
      }
      if (!isControlled) {
        setChecked(!checkedState);
      }
      if (onClick) {
        onClick(e);
      }
    },
    [checkedState, disabled, isControlled, onClick],
  );

  if (isControlled && props.checked !== checkedState) {
    setChecked(props.checked!);
  }

  return {
    checkedState,
    handleChange,
    handleClick,
    setChecked,
  };
}
