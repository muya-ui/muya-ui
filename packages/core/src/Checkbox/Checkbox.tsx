import React, { forwardRef, Ref, useMemo } from 'react';

import { CheckboxIcon, CheckIcon, HalfcheckboxIcon } from '@muya-ui/theme-light';

import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import {
  StyledCheckboxContent,
  StyledCheckboxIconWrap,
  StyledCheckboxInput,
  StyledCheckboxLabel,
  StyledCheckboxSpan,
} from './styled';
import { ICheckboxProps } from './types';
import useCheckbox from './useCheckbox';

const defaultStyles = {
  wrapper: '',
  icon: '',
  iconWrap: '',
  span: '',
};

const Checkbox = forwardRef(function Checkbox(props: ICheckboxProps, ref: Ref<HTMLLabelElement>) {
  const {
    checked,
    children,
    defaultChecked = false,
    disabled = false,
    indeterminate = false,
    autoFocus = false,
    readOnly = false,
    onChange,
    onMouseEnter,
    onMouseLeave,
    size = 'm',
    styles,
    style,
    className,
    inputRef,
    ellipsis = false,
    width,
    ...restProps
  } = props;
  const theme = useTheme();
  const innerStyles = useStyles('checkbox', defaultStyles, styles);
  const { checkedState, handleChange } = useCheckbox(props, 'checked' in props);
  const icons = useMemo(
    () => ({
      checked: CheckboxIcon,
      unChecked: CheckIcon,
      indeterminate: HalfcheckboxIcon,
      ...theme.components.Checkbox.icons,
    }),
    [theme.components.Checkbox.icons],
  );
  const {
    indeterminate: IndeterminateIcon,
    checked: CheckedIcon,
    unChecked: UnCheckedIcon,
  } = icons;
  const { iconBgColor, iconColor } = theme.components.Checkbox;
  const iconNode = useMemo(() => {
    if (indeterminate) {
      return <IndeterminateIcon bgColor={iconBgColor} {...innerStyles.icon} />;
    }
    if (checkedState) {
      return <CheckedIcon bgColor={iconBgColor} {...innerStyles.icon} />;
    }
    return (
      <UnCheckedIcon
        bgColor={iconBgColor}
        color={iconColor.unChecked.normal}
        {...innerStyles.icon}
      />
    );
  }, [checkedState, iconBgColor, iconColor.unChecked.normal, indeterminate, innerStyles.icon]);

  return (
    <StyledCheckboxLabel
      $disabled={disabled}
      $size={size}
      theme={theme}
      $width={width}
      $ellipsis={ellipsis}
      $checked={checkedState}
      $indeterminate={indeterminate!}
      ref={ref}
      className={[innerStyles.wrapper.className, className].join(' ').trim()}
      style={{ ...innerStyles.wrapper.style, ...style }}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      <StyledCheckboxSpan {...innerStyles.span} onClick={e => e.stopPropagation()}>
        <StyledCheckboxInput
          {...restProps}
          autoFocus={autoFocus}
          readOnly={readOnly}
          ref={inputRef}
          type="checkbox"
          disabled={disabled}
          checked={checkedState}
          theme={theme}
          onChange={handleChange}
        />
        <StyledCheckboxIconWrap theme={theme} {...innerStyles.iconWrap}>
          {iconNode}
        </StyledCheckboxIconWrap>
      </StyledCheckboxSpan>
      {children ? <StyledCheckboxContent theme={theme}>{children}</StyledCheckboxContent> : null}
    </StyledCheckboxLabel>
  );
});
export default React.memo(Checkbox);
