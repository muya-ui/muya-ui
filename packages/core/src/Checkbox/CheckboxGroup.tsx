import React, { Ref, useMemo } from 'react';

import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import Checkbox from './Checkbox';
import { StyledCheckboxGroup } from './styled';
import { ICheckboxGroupOption, ICheckboxGroupProps } from './types';
import useCheckboxGroup from './useCheckboxGroup';

const CheckboxGroup = memoForwardRef(function Group(
  props: ICheckboxGroupProps,
  ref: Ref<HTMLDivElement>,
) {
  const { size = 'm', styles } = props;
  const {
    opts,
    restProps,
    valueState,
    children,
    renderCheckbox,
    checkboxEllipsis,
    checkboxWidth,
    handleChange,
  } = useCheckboxGroup(props);
  const theme = useTheme();
  const checkboxList = useMemo(() => {
    if (opts && opts.length) {
      return opts.map((option, optIndex) => {
        const { label, value, disabled, ellipsis, width } = option;
        const checkboxNode = (
          <Checkbox
            key={value.toString()}
            styles={styles}
            size={size}
            value={value}
            ellipsis={ellipsis === undefined ? checkboxEllipsis : ellipsis}
            width={width || checkboxWidth}
            disabled={props.disabled || disabled}
            checked={valueState.indexOf(value) > -1}
            onChange={handleChange.bind(null, option)}
          >
            {label}
          </Checkbox>
        );
        return renderCheckbox(checkboxNode, option, optIndex);
      });
    }

    return React.Children.map(children, (child, childIndex) => {
      if (!React.isValidElement(child)) {
        return null;
      }
      const {
        value,
        disabled,
        onChange,
        ellipsis,
        width,
        children: label,
        ...otherChildProps
      } = child.props;
      const option: ICheckboxGroupOption = { value, disabled, onChange, label };
      const checkboxNode = React.cloneElement(child, {
        ...otherChildProps,
        children: label,
        value,
        ellipsis: ellipsis === undefined ? checkboxEllipsis : ellipsis,
        width: width || checkboxWidth,
        checked: valueState.indexOf(value) > -1,
        disabled: props.disabled || disabled,
        styles,
        size,
        onChange: handleChange.bind(null, option),
      });
      return renderCheckbox(checkboxNode, option, childIndex);
    });
  }, [
    checkboxEllipsis,
    checkboxWidth,
    children,
    handleChange,
    opts,
    props.disabled,
    renderCheckbox,
    size,
    styles,
    valueState,
  ]);

  return (
    <StyledCheckboxGroup {...restProps} ref={ref} theme={theme}>
      {checkboxList}
    </StyledCheckboxGroup>
  );
});

export default CheckboxGroup;
