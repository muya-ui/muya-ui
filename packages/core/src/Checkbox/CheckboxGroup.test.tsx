import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import { ICheckboxGroupValue } from '@muya-ui/core';

import Checkbox from './Checkbox';
import CheckboxGroup from './CheckboxGroup';

test('测试字符串Option', () => {
  const options = ['a', 'b', 'c'];
  let list: ICheckboxGroupValue[] = [];
  const wrap = mount(
    <CheckboxGroup
      options={options}
      value={list}
      onChange={value => {
        list = value;
      }}
    />,
  );
  wrap
    .find('input')
    .first()
    .simulate('change', { target: { checked: true } });
  expect(list[0]).toEqual('a');
});

test('测试 value 为 undefined', () => {
  const options = ['a', 'b', 'c'];
  let list: ICheckboxGroupValue[] = [];
  const wrap = mount(
    <CheckboxGroup
      options={options}
      value={undefined}
      onChange={value => {
        list = value;
      }}
    />,
  );
  wrap
    .find('input')
    .first()
    .simulate('change', { target: { checked: true } });
  expect(list[0]).toEqual('a');
});

test('测试对象Option', () => {
  const onChange1 = sinon.stub();
  const onChange2 = sinon.stub();
  const options = [
    { label: 'a', value: 1, onChange: onChange1 },
    { label: 'b', value: 2, onChange: onChange2 },
    { label: 'c', value: 3 },
    { label: 'd', value: 4, disabled: true },
  ];

  const wrap = mount(<CheckboxGroup options={options} defaultValue={[2]} />);
  expect(
    wrap
      .find('div')
      .childAt(1)
      .find('input')
      .prop('checked'),
  ).toEqual(true);
  wrap
    .find('input')
    .first()
    .simulate('change', { target: { checked: true } });
  expect(() => {
    sinon.assert.calledOnce(onChange1);
    sinon.assert.notCalled(onChange2);
  }).not.toThrow();
  wrap
    .find('div')
    .childAt(1)
    .find('input')
    .simulate('change', { target: { checked: false } });
  expect(() => {
    sinon.assert.calledOnce(onChange2);
  }).not.toThrow();
});

test('测试没有任何参数', () => {
  const tree = renderer.create(<CheckboxGroup />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试自定义渲染 CheckBox', () => {
  const tree = renderer
    .create(
      <CheckboxGroup renderCheckbox={node => <div>{node}</div>}>
        111
        <Checkbox value={1}>1</Checkbox>
        <Checkbox value={2}>2</Checkbox>
        <Checkbox value={3}>3</Checkbox>
      </CheckboxGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试自定义渲染 CheckBox width ellipsis', () => {
  const tree = renderer
    .create(
      <CheckboxGroup checkboxEllipsis checkboxWidth={100}>
        111
        <Checkbox value={1}>1</Checkbox>
        <Checkbox value={2}>2</Checkbox>
        <Checkbox value={3}>3</Checkbox>
      </CheckboxGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
