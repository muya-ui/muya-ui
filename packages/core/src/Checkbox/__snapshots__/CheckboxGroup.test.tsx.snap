// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试没有任何参数 1`] = `
.c0 .c1 {
  margin-right: 8px;
}

<div
  className="c0"
/>
`;

exports[`测试自定义渲染 CheckBox 1`] = `
.c9 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  color: rgba(204,204,204,1);
  fill: currentColor;
}

.c6 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  cursor: pointer;
  opacity: 0;
}

.c4 {
  position: relative;
  top: -0.09em;
  display: inline-block;
  line-height: 1;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  cursor: pointer;
}

.c2 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  display: inline-block;
  font-size: 12px;
  line-height: 0;
  font-variant: tabular-nums;
  font-feature-settings: 'tnum';
  color: rgba(51,51,51,1);
  cursor: pointer;
}

.c2 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c2 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(204,204,204,1);
}

.c2 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c2:hover .c7 .c8 {
  color: rgba(26,122,248,1);
}

.c2:active .c7 .c8 {
  color: rgba(12,108,235,1);
}

.c11 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c11 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(26,122,248,1);
}

.c11 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c11:hover .c7 .c8 {
  color: rgba(51,139,255,1);
}

.c11:active .c7 .c8 {
  color: rgba(12,108,235,1);
}

.c12 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c12 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(204,204,204,1);
}

.c12 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c12 .c5,
.c12 .c3 {
  cursor: not-allowed;
}

.c0 .c1 {
  margin-right: 8px;
}

<div
  className="c0"
>
  <div>
    <label
      className="c1 c2 muya-checkbox-wrapper"
      style={Object {}}
    >
      <span
        className="c3 c4 muya-checkbox-span"
        onClick={[Function]}
      >
        <input
          autoFocus={false}
          checked={false}
          className="c5 c6"
          disabled={false}
          onChange={[Function]}
          readOnly={false}
          type="checkbox"
          value={1}
        />
        <span
          className="c7 muya-checkbox-icon-wrap"
        >
          <svg
            aria-hidden="false"
            className="c8 c9 muya-checkbox-icon"
            color="rgba(204,204,204,1)"
            focusable="false"
            viewBox="0 0 1024 1024"
          >
            <rect
              fill="rgba(255,255,255,1)"
              height={716.8}
              width={716.8}
              x={153.60000000000002}
              y={153.60000000000002}
            />
            <path
              d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
            />
          </svg>
        </span>
      </span>
      <span
        className="c10 "
      >
        1
      </span>
    </label>
  </div>
  <div>
    <label
      className="c1 c2 muya-checkbox-wrapper"
      style={Object {}}
    >
      <span
        className="c3 c4 muya-checkbox-span"
        onClick={[Function]}
      >
        <input
          autoFocus={false}
          checked={false}
          className="c5 c6"
          disabled={false}
          onChange={[Function]}
          readOnly={false}
          type="checkbox"
          value={2}
        />
        <span
          className="c7 muya-checkbox-icon-wrap"
        >
          <svg
            aria-hidden="false"
            className="c8 c9 muya-checkbox-icon"
            color="rgba(204,204,204,1)"
            focusable="false"
            viewBox="0 0 1024 1024"
          >
            <rect
              fill="rgba(255,255,255,1)"
              height={716.8}
              width={716.8}
              x={153.60000000000002}
              y={153.60000000000002}
            />
            <path
              d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
            />
          </svg>
        </span>
      </span>
      <span
        className="c10 "
      >
        2
      </span>
    </label>
  </div>
  <div>
    <label
      className="c1 c2 muya-checkbox-wrapper"
      style={Object {}}
    >
      <span
        className="c3 c4 muya-checkbox-span"
        onClick={[Function]}
      >
        <input
          autoFocus={false}
          checked={false}
          className="c5 c6"
          disabled={false}
          onChange={[Function]}
          readOnly={false}
          type="checkbox"
          value={3}
        />
        <span
          className="c7 muya-checkbox-icon-wrap"
        >
          <svg
            aria-hidden="false"
            className="c8 c9 muya-checkbox-icon"
            color="rgba(204,204,204,1)"
            focusable="false"
            viewBox="0 0 1024 1024"
          >
            <rect
              fill="rgba(255,255,255,1)"
              height={716.8}
              width={716.8}
              x={153.60000000000002}
              y={153.60000000000002}
            />
            <path
              d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
            />
          </svg>
        </span>
      </span>
      <span
        className="c10 "
      >
        3
      </span>
    </label>
  </div>
</div>
`;

exports[`测试自定义渲染 CheckBox width ellipsis 1`] = `
.c9 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  color: rgba(204,204,204,1);
  fill: currentColor;
}

.c6 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  cursor: pointer;
  opacity: 0;
}

.c4 {
  position: relative;
  top: -0.09em;
  display: inline-block;
  line-height: 1;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  cursor: pointer;
}

.c11 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c11 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(204,204,204,1);
}

.c11 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c11:hover .c7 .c8 {
  color: rgba(26,122,248,1);
}

.c11:active .c7 .c8 {
  color: rgba(12,108,235,1);
}

.c12 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c12 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(26,122,248,1);
}

.c12 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c12:hover .c7 .c8 {
  color: rgba(51,139,255,1);
}

.c12:active .c7 .c8 {
  color: rgba(12,108,235,1);
}

.c13 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c13 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(204,204,204,1);
}

.c13 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c13 .c5,
.c13 .c3 {
  cursor: not-allowed;
}

.c2 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  display: inline-block;
  font-size: 12px;
  line-height: 0;
  font-variant: tabular-nums;
  font-feature-settings: 'tnum';
  color: rgba(51,51,51,1);
  cursor: pointer;
  width: 100px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 .c10 {
  padding: 0 4px;
  line-height: 16px;
}

.c2 .c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-top;
  font-size: 16px;
  height: 16px;
  width: 16px;
  line-height: 16px;
  padding: 0;
  color: rgba(204,204,204,1);
}

.c2 .c7 .c8 {
  z-index: 1;
  -webkit-transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
  transition: color cubic-bezier(0.4,0,0.2,1) 150ms;
}

.c2:hover .c7 .c8 {
  color: rgba(26,122,248,1);
}

.c2:active .c7 .c8 {
  color: rgba(12,108,235,1);
}

.c0 .c1 {
  margin-right: 8px;
}

<div
  className="c0"
>
  <label
    className="c1 c2 muya-checkbox-wrapper"
    style={Object {}}
  >
    <span
      className="c3 c4 muya-checkbox-span"
      onClick={[Function]}
    >
      <input
        autoFocus={false}
        checked={false}
        className="c5 c6"
        disabled={false}
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value={1}
      />
      <span
        className="c7 muya-checkbox-icon-wrap"
      >
        <svg
          aria-hidden="false"
          className="c8 c9 muya-checkbox-icon"
          color="rgba(204,204,204,1)"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
          />
        </svg>
      </span>
    </span>
    <span
      className="c10 "
    >
      1
    </span>
  </label>
  <label
    className="c1 c2 muya-checkbox-wrapper"
    style={Object {}}
  >
    <span
      className="c3 c4 muya-checkbox-span"
      onClick={[Function]}
    >
      <input
        autoFocus={false}
        checked={false}
        className="c5 c6"
        disabled={false}
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value={2}
      />
      <span
        className="c7 muya-checkbox-icon-wrap"
      >
        <svg
          aria-hidden="false"
          className="c8 c9 muya-checkbox-icon"
          color="rgba(204,204,204,1)"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
          />
        </svg>
      </span>
    </span>
    <span
      className="c10 "
    >
      2
    </span>
  </label>
  <label
    className="c1 c2 muya-checkbox-wrapper"
    style={Object {}}
  >
    <span
      className="c3 c4 muya-checkbox-span"
      onClick={[Function]}
    >
      <input
        autoFocus={false}
        checked={false}
        className="c5 c6"
        disabled={false}
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value={3}
      />
      <span
        className="c7 muya-checkbox-icon-wrap"
      >
        <svg
          aria-hidden="false"
          className="c8 c9 muya-checkbox-icon"
          color="rgba(204,204,204,1)"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M768 42.667c117.824 0 213.333 95.509 213.333 213.333v512c0 117.824-95.509 213.333-213.333 213.333H256c-117.824 0-213.333-95.509-213.333-213.333V256c0-117.824 95.509-213.333 213.333-213.333h512zm0 64H256c-81.365 0-147.52 65.066-149.301 145.994l-.032 3.339v512c0 81.365 65.066 147.52 145.994 149.301l3.339.032h512c81.365 0 147.52-65.066 149.301-145.994l.032-3.339V256c0-81.365-65.066-147.52-145.994-149.301l-3.339-.032z"
          />
        </svg>
      </span>
    </span>
    <span
      className="c10 "
    >
      3
    </span>
  </label>
</div>
`;
